Mini proyecto de regresion lineal | Examen Medio Curso | Redes Neuronales Artificiales

Autor: Alex Garcia Serna
Matricula:  1725239
Correo: alex.garciase@uanl.edu.mx
Github: github.com/AlexGarSer

Objetivo:
El mini proyecto consiste en desarrollar un algoritmo que recibe como entrada valores dados por el usuario de las variables de control seleccionadas como relevantes y devuelve al usuario el tiempo de producción esperado si se ajustan los controles con dichos valores dados.

La informacion utilizada para este proyecto al ser de caracter sensible se eliminaron los nombres de las columnas por razones de confidencialidad y se sustituyeron por "X" + "# de la columna".

In [94]:
# El dataset utilizado en este proyecto se encuentra nombrado como ProblemaRegresion.csv
# Este dataset fue otorgado por el instructor y para uso exclusivo del examen.

#Librerias a utlizar
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
#Primero se leen los datos del dataset y los pasamos a un dataframe para poder manipularlos.
datos = pd.read_csv("ProblemaRegresion.csv")

Al darnos cuenta que faltaba informacion en el dataset y marcaba errrores al querer utilizarla decidimos utilizar los datos de media que teniamos actualmente para rellenar los espacios vacios.

In [95]:
#Se imprime la media de todas las columnas del dataframe
print(datos.mean())
print('\n'*3)

y       36.519935
x1      28.651091
x2      53.725729
x3      17.623180
x4      40.693215
x5      16.716437
x6      23.954144
x7       1.453853
x8     184.806763
x9     146.621576
x10    125.241155
x11    120.654717
x12     59.565670
x13    477.957222
x14     40.567198
x15     56.332647
x16      1.797769
x17     93.638739
x18      3.841009
x19     84.598280
x20     18.863547
x21      5.701706
x22     94.396895
x23      4.080073
x24     85.413006
x25      7.146739
x26     23.322593
x27     93.797414
x28      3.703272
x29     84.278472
x30     92.663914
x31      3.439937
x32     83.958096
x33     13.799972
x34      4.036723
dtype: float64






Una vez completados todos los datos, procedo a seccionar las salidas y las caracteristicas del dataset
donde Y representa las salidas esperadas y X son las caracteristicas

In [96]:
Y=datos[['y']] # se filtran la columna de salida esperada
print(Y[:5]) # Se imprime para corroborar que sean los datos esperados

print('\n'*3)

# Se filtran las columans que conformaran las caracteristicas
X = datos[['x1','x2','x4','x5','x6','x7','x8','x9','x10','x11','x12','x14','x15','x16','x17','x18','x19','x20','x21','x22','x24','x25','x26','x27','x28','x29','x30','x31','x32','x34']] #Caracteristicas del conjunto
print(X[:5]) # Se imprime para corroborar que sean los datos esperados


           y
0  31.303077
1  32.520000
2  31.936000
3  31.822500
4  31.383636




          x1         x2         x4         x5         x6        x7  \
0  23.403665  61.001078  37.547146  16.806608  24.772692  1.497338   
1  20.864591  65.232089  39.407843  16.170943  23.027157  1.425443   
2  21.788283  63.692885  40.569120  15.907050  23.273360  1.470490   
3  24.315423  59.481762  39.062450  16.688242  23.917008  1.448125   
4  24.649930  58.924352  39.968936  16.013227  23.566355  1.483309   

           x8          x9         x10         x11  ...   x24       x25  \
0  171.359231  144.714615  114.259231  126.846154  ...  86.1  7.237142   
1  176.830000  143.911429  119.740000  124.571429  ...  86.1  5.811863   
2  176.234000  146.550000  118.956000  124.300000  ...  85.6  6.908267   
3  174.932500  147.945833  116.710000  126.750000  ...  85.6  0.931507   
4  176.130000  147.524545  118.110909  127.545455  ...  85.6  4.306220   

         x26   x27  x28   x29      x30    x31      x

In [97]:
#Se imprime el dataframe
print(datos)
print('\n'*3)

             y         x1         x2         x3         x4         x5  \
0    31.303077  23.403665  61.001078  15.595257  37.547146  16.806608   
1    32.520000  20.864591  65.232089  13.903320  39.407843  16.170943   
2    31.936000  21.788283  63.692885  14.518832  40.569120  15.907050   
3    31.822500  24.315423  59.481762  16.202815  39.062450  16.688242   
4    31.383636  24.649930  58.924352  16.425718  39.968936  16.013227   
..         ...        ...        ...        ...        ...        ...   
761  35.261250  26.110933  56.618658  17.270409  40.162700  17.288412   
762  34.576000  24.906090  58.625029  16.468882  42.486990  16.380330   
763  36.841000  26.603597  55.798247  17.598156  42.065450  16.348790   
764  36.717500  16.846006  72.047118  11.106877  42.785300  16.426838   
765  36.519935  28.651091  53.725729  17.623180  40.693215  16.716437   

            x6        x7          x8          x9  ...       x25        x26  \
0    24.772692  1.497338  171.359231  144.714

In [98]:
#Se imprime la cantidad de filas y columnas que tiene el datasframe
print(datos.shape)
print('\n'*3)

(766, 35)






Para empezar a trabajar con los datos primero los pasaremos por una transformacion ya que algunos de los datos no son de las mismas magnitudes los pasaremos por una transformacion de centrado, escalado y normalizado, para ellos utilizaremos la libreria Sklearn con su funcion preprocessing.StandardScaler

In [99]:
scaler = preprocessing.StandardScaler()
datos=scaler.fit_transform(X)
print(scaler)

StandardScaler()


Una vez que ya tenemos los datos es hora de seccionar cuales de ellos seran utilizados para el entrenamiento de la red neuronal y cuales seran la comprobacion de que la red neuronal predice de manera correcta.

En este caso utilizaremos una proporcion 70%:30% para otorgar mas ejemplos de entrenamiento a la red neuronal que la que va a predecir.

Al ser 766 datos utilizaremos 536 que es el equivalente al 70% de los datos para el conjunto de entrenamiento y solo 230 para el conjunto de pruebas para corroborar el funcionamiento de la red neuronal.

In [100]:
#Division en conjunto de entrenamiento y conjunto de prueba
x_train, x_test, y_train, y_test = train_test_split(X,Y,test_size=0.3)

# Se imprime la cantidad de pruebas por cada set
print(x_train.shape) #Se imprime la cantidade columnas y filas a utilizar
print(x_test.shape) #Se imprime la cantidade columnas y filas a utilizar
print("\n"*3)

#Estandar de conjuntos de entrenamiento y prueba
x_train = StandardScaler().fit_transform(x_train.astype(str))
x_test = StandardScaler().fit_transform(x_test)

print("Estandarizando...\n"*3)
print("Estandarizado Completado!")

(536, 30)
(230, 30)




Estandarizando...
Estandarizando...
Estandarizando...

Estandarizado Completado!


In [101]:

print("Entrenamiento")
print(x_train[:5])
print("Prueba")
print(x_test[:5])

Entrenamiento
[[-0.25492477  0.41534828  0.20156949 -0.37109413  0.4000494   0.68966529
  -0.17835415 -0.11894139  0.00455405  0.30807198 -0.21950952  0.8557226
   0.41295391 -0.43819665  0.4304079   0.14314863  0.15985808 -0.14157252
  -0.11864886  0.66155513 -0.10410949 -0.21528953 -0.04817969 -0.64985342
  -0.77963713 -0.16316742  1.0069816  -0.7418685   0.82652615 -0.05649251]
 [-0.26135566  0.55369593  1.47532453 -1.31987829 -1.29069902 -0.13321737
   0.96055578 -0.40981444  0.9530222   0.65862541  0.0234937   1.26002048
   1.35445448 -0.46703145 -0.44818656  0.75223445 -0.07181624  0.5624545
   1.61257108 -1.54490398 -0.90416046  1.5100603   0.58442486 -0.1709935
   0.69952295 -0.76209395 -0.23082534  0.11646609  2.12733984 -0.34659889]
 [-0.41180498  0.40611842 -0.2652248   0.47234347  0.14148952 -0.3532453
  -0.11601473 -0.60737652 -0.31022589  0.10977913  0.22840778 -0.35026311
   0.53005397 -0.51687875  0.30214011  0.29037951 -0.01127946 -0.57402991
  -0.54087906 -0.0099759  

Continuaremos aplicando ahora si la regresion lineal utilizando los datos de entrenamiento

In [ ]:
regresion = linear_model.LinearRegression()
regresion.fit(x_train,y_train)